### PR TITLE
containers: Move unit-tests to Debian stable, self-validate changes

### PR DESCRIPTION
**Pull request by martinpitt**

### CODE REVIEW

**Comment by martinpitt** [comment]
Not bad, not bad! Just the [i386 n install failure](https://github.com/cockpit-project/cockpit/pull/14852/checks?check_run_id=1343756948) needs fixing. This worked better with the distro package, but `node-sass` fails to `npm install` with it, as there is no compiled i386 version for that any more. I'll try something else (installing nodejs:amd64 on i386 as well, essentially what the `n` stuff does right now), and if that does not work, fix it with `n`.

**Comment by martinpitt** [comment]
I'm still thinking how the really cool kids would extend this to automatically upload the container to dockerhub. The cleanest way would be to use the https://hub.docker.com/u/cockpitbuilder user, create a token for it, and stuff that into our org secrets. But I don't have access to that dockerhub user, and I don't think we have actually used it for anything in years; it *is* a member of the cockpit org builders team, though. There is a https://hub.docker.com/u/cockpituous user which I *can* login as, but it doesn't have any privileges nor history.

@stefwalter, @larskarlitski , @mvollmer, do you know the story behind cockpituous vs. cockpitbuilder, and do you have the credentials for the latter? Supposedly we should remove one of them and use the other for automatic container builds?

**Comment by martinpitt** [comment]
All green :balloon: ready for review. Once this lands, I'll rebuild and push the container manually, and then I'll work on regularly refreshing that (gated on tests, of course) every week, so that we don't fall into this trap again. For that I'd like to sort out the bot user story, though.

**Comment by larskarlitski** [comment]
> @stefwalter, @larskarlitski , @mvollmer, do you know the story behind cockpituous vs. cockpitbuilder, and do you have the credentials for the latter? Supposedly we should remove one of them and use the other for automatic container builds?

No idea, sorry. I don't have credentials.

**Comment by stefwalter** [comment]
I looked through my email and saved passwords, and can't find reference to the cockpitbuilder user. I must have known what it was once upon a time, but that knowledge has now escaped me :S

**Comment by martinpitt** [comment]
Thanks @stefwalter, @mvollmer, and @larskarlitski ! As nobody is able to log in as cockpitbuilder any more, let's just declare it dead then. I instead added the cockpituous user to our organization's builder team. I committed the bits into our ci-secrets git.

File: .github/workflows/unit-tests.yml [hunk 1]
```diff
@@ -40,5 +50,10 @@ jobs:
           # need this to also fetch tags
           fetch-depth: 0
 
+      - name: Build unit test container if it changed
```
Comment (allisonkarlitskaya) [inline .github/workflows/unit-tests.yml:53]: Can we DRY this a bit somehow?

File: containers/unit-tests/Dockerfile [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
-FROM ${debian_arch}/ubuntu:19.04
+FROM ${debian_arch}/debian:stable
 
 ARG personality=linux64
 COPY setup.sh /
 RUN ${personality} /setup.sh ${personality} && rm -rf /setup.sh
 
 # HACK: Working around Node.js screwing around with stdio
-ENV NODE_PATH /usr/local/bin/node.lts
+ENV NODE_PATH /usr/bin/node
```
Comment (allisonkarlitskaya) [inline containers/unit-tests/Dockerfile:9]: yay!

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]

File: containers/unit-tests/setup.sh [hunk 3]
```diff
@@ -7,9 +7,10 @@ dependencies="\
     autoconf \
     automake \
     build-essential \
-    chromium-browser \
-    clang python3 \
+    chromium \
+    clang \
     curl \
+    debian-archive-keyring \
```
Comment (allisonkarlitskaya) [inline containers/unit-tests/setup.sh:13]: Why is this necessary?  amd64?  debug?  both?

We install this only as part of `${dependencies}`, which is the last thing we install, so it's only installed after it might have been useful... Is this really necessary at all, or did I miss something?

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]

File: containers/unit-tests/setup.sh (hunk 3 above)
Comment (martinpitt) [reply containers/unit-tests/setup.sh:13]: Well spotted, this is a leftover from the "move to ubuntu 20.04" variant, which needed to pull chromium from Debian. Removed.

File: .github/workflows/unit-tests.yml (hunk 1 above)
Comment (martinpitt) [reply .github/workflows/unit-tests.yml:53]: I added a new commit to make the build command shorter, which helps a bit already.

Beyond that, it would be nice to factorize/parameterize the whole `steps:` or even the whole `job` entry. I have not yet learned about [github workflow custom actions](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions) to do that. I want to do that when introducing automatic unit-test container refresh with gating, as that will have to run through similar steps -- but it will be *more* code in total, and thus also be harder to understand what happens at a glance, so I'd like to not commit any of that just yet.

So, I know that this is repetitive, but IMHO it doesn't hurt too much for now.

**Comment by martinpitt** [comment]
Awesome, now they all [fail](https://github.com/cockpit-project/cockpit/pull/14852/checks?check_run_id=1351260825) like
```
make[2]: Leaving directory '/tmp/source'
make[2]: write error: stdout
```

That sounds like a turd-polish thing again. However, my recent force pushes don't change anything about invoking the test, just building the container..

**Comment by martinpitt** [comment]
Going back to the previous version that was green, just to compare if it's something in the base image or npm module versions that caused this breakage.

**Comment by martinpitt** [comment]
Oooh, I know what is wrong -- `ubuntu-latest` has podman, so the [current run](https://github.com/cockpit-project/cockpit/actions/runs/345258268) builds a *podman* container, and then *runs* the tests in docker. Thus the
```diff
-          [ -z "$changes" ] || docker build --build-arg debian_arch=amd64 --build-arg personality=linux64 -t docker.io/cockpit/unit-tests containers/unit-tests
+          [ -z "$changes" ] || containers/unit-tests/build amd64

```
change was the root cause. So mystery solved, but we still have that annoying "stdout: write failed" bug that has blocked the container update for > 1 year..

**Comment by martinpitt** [comment]
So if printing a > 64 KiB tar command line does not work somewhere, then ... let's just not. I added a commit to quiesce the build. In my local test it gets past that "write error: stdout" failure now.

**Comment by martinpitt** [comment]
Locally it now fails much later, with this:
```
Traceback (most recent call last):
  File "../../tools/tap-driver", line 416, in <module>
    sys.exit(main(sys.argv))
  File "../../tools/tap-driver", line 413, in main
    return driver.run()
  File "../../tools/tap-driver", line 263, in run
    returncode = self.execute()
  File "../../tools/tap-driver", line 171, in execute
    assert fcntl.fcntl(1, fcntl.F_GETFL) & os.O_NONBLOCK == 0
AssertionError
```

So it seems we are missing some spot where node messes up stdout :sob: . Let's see how it fares on gh.

**Comment by martinpitt** [comment]
Funny.. Today I thought "what if we pretended that all is well", and did a run with *all* of our hacks removed. And voilà, I get a clean run both locally and on CI! Opinions?

**Comment by martinpitt** [comment]
Meh, the [i386 run](https://github.com/cockpit-project/cockpit/pull/14852/checks?check_run_id=1374118834) got stuck on a hanging test -- at least *that* weird bug didn't magically go away :sob: But retrying it will give us more data about how robust the tests are wrt. the "stdout: write error" thing.

File: containers/unit-tests/build [hunk 4]
```diff
@@ -7,7 +8,10 @@ if test -z "${docker:=$(which podman || which docker || true)}"; then
   exit 1
 fi
 
-$docker build --build-arg debian_arch=amd64 --build-arg personality=linux64 -t cockpit/unit-tests ${dir}
-$docker tag  cockpit/unit-tests:latest docker.io/cockpit/unit-tests:latest
-$docker build --build-arg debian_arch=i386 --build-arg personality=linux32 -t cockpit/unit-tests:i386 ${dir}
-$docker tag  cockpit/unit-tests:i386 docker.io/cockpit/unit-tests:i386
+if [ -z "${1:-}" ] || [ "${1:-}" = amd64 ]; then
```
Comment (allisonkarlitskaya) [inline containers/unit-tests/build:11]: You could simplify that to `[ "${1:-amd64}" = amd64 ]`.  Arguably less readable, though.

File: tools/tap-driver [hunk 5]
```diff
@@ -166,11 +166,6 @@ class Driver:
 
         proc.wait()
 
-        # Make sure the test didn't change blocking output
-        assert fcntl.fcntl(0, fcntl.F_GETFL) & os.O_NONBLOCK == 0
```
Comment (allisonkarlitskaya) [inline tools/tap-driver:170]: I have the feeling this will bite us again in the future.  Maybe github is just faster about emptying its pipes than semaphore was, and that's why it doesn't break anymore?

I'd like to find a more robust solution here...

If it works, though, I don't block[1] it.

[1] :roll_eyes:

**Review by allisonkarlitskaya — APPROVED** [review]
